### ISSUE TITLE
Super high latency when using Cloud Spanner Go client

### ISSUE THREAD

**Issue by ethanyanjiali** [GoogleCloudPlatform/google-cloud-go#1207]
(delete this for feature requests)
## Client

Cloud Spanner

## Describe Your Environment

App Engine Standard

## Expected Behavior

Query takes similar time as the Spanner Cloud Console shows

## Actual Behavior

Query using Go client is extremely slower than what Spanner Cloud Console shows
For example, a simple read takes 2ms on the console, but when I use go client and time it, it takes like 1000ms.

**Comment by enocom** [comment]
How are you timing the Go client?

**Comment by ethanyanjiali** [comment]
@enocom Thanks for following up on this. I just logged the timestamp. Here's my sample code.
```go
start := time.Now().UnixNano() 
log.Infof(ctx, "start %v", start) // start 1541372927986170088
stmt := cloudspanner.NewStatement("SELECT * FROM TableA")
it := s.client.Single().Query(ctx, stmt)
log.Infof(ctx, "make single query %v", time.Now().UnixNano()-start) // make single query 1199093250ns ~ 1199ms
```
Another example is:
```go
start := time.Now().UnixNano()
log.Infof(ctx, "start %v", start) // start 1541372952689569764
m, err := cloudspanner.InsertStruct(TableAId, entityToInsert)
if err != nil {
    return err
}
log.Infof(ctx, "create mutation %v", time.Now().UnixNano()-start) // create mutation 16312ns
_, err = s.client.Apply(ctx, []*cloudspanner.Mutation{m})
if err != nil {
    return err
}
log.Infof(ctx, "apply mutation %v", time.Now().UnixNano()-start) // apply mutation 1668889188ns ~ 1668ms
```
I tested this on a F4_1G GAE instance. i know the instance speed and network could affect the benchmark here, but i thought GAE has already optimize the request between GAE instance and Spanner instance. And also F4_1G shouldn't be a slow instance. So that's why I'm suspecting the spanner Go client here. Do you know what could be the possible way to reduce the latency to something under 200ms? Thanks.

**Comment by enocom** [comment]
Thanks for the detailed follow-up. Are these high values you're seeing consistent, or are they the outliers?

**Comment by ethanyanjiali** [comment]
@enocom yes they are consistent, sometimes even fluctuate to even higher. Also when I test locally, I actually got even lower latency:
```
2018/11/06 17:04:44 INFO: start 1541523884879125787
2018/11/06 17:04:45 INFO: make single query 975866299  (975ms)
```
I thought the connect between my localhost and spanner instance should be slower. We have some firewall rules set up on the GAE project. Don't think that would slow down the internal db connect though.

Could it come from some GAE sandbox configuration? Because I've noticed one more thing, not sure if it's related: When I use big query Go client, the latency also increased around 1s. For example, a query complete in 1.2s on cloud console, but I got 2.2s on my GAE instance. However we don't expect BQ to be fast so that doesn't bother us too much so far.

**Comment by ethanyanjiali** [comment]
@enocom some further observation:
So I switched to a compute engine instance `n1-standard-1 (1 vCPU, 3.75 GB memory)`, and the latency goes down to 400ms for both spanner read/write query. also for big query it goes down for 1s as well. it seems like App Engine is the reason. Not sure what configuration on GAE caused this though.

Is there any other way to further lower the Spanner latency? Still have no idea how would it possible to reach 3ms for the read query like the console says. I'm just using 1 spanner node for development now but I doubt that has effect on latency.

**Comment by enocom** [comment]
@broady @sbuss What do you two think about this? Are there good ways to lower the latency for connections between App Engine and Spanner?

**Comment by broady** [comment]
Can you clarify if you're using the new Go 1.11 beta runtime?

If you're not, you're using [GAE's Socket API](https://cloud.google.com/appengine/docs/standard/go/sockets/), which actually is not covered by any SLA.

The new beta runtime should improve network performance to Spanner, but I think there are still some issues with regard to how packets are routed between GAE and cloud services like Spanner.

**Comment by ethanyanjiali** [comment]
@broady Thanks for looking into it. We are indeed using App Engine Standard with Go 1.8 not 1.11, didn't know it could cause latency issue

**Comment by ethanyanjiali** [comment]
Using Flex with Go 1.11 doesn't lower the latency though, the only possible workaround for me is to migrate to GKE or GCE

**Comment by broady** [comment]
I wasn't suggesting Flex, but it's very odd the latency isn't good from Flex.

Can you confirm your Spanner instance is in the same zone (or region) as your Flex instance?

Flex is a relatively thin layer on top of GCE, so you shouldn't get better performance going to GKE/GCE.

**Comment by ethanyanjiali** [comment]
Ah I think that's the reason. I mistakenly put Spanner (us-west1) and App Engine (us-central) in different regions. And now I migrate Spanner to us-central1 just as App Engine it lowers to 400ms. Is that the lowest latency I can get from the combination of GAE and Spanner?

**Comment by odeke-em** [comment]
Hello @ethanyanjiali 
>So I switched to a compute engine instance n1-standard-1 (1 vCPU, 3.75 GB memory), and the latency goes down to 400ms for both spanner read/write query. also for big query it goes down for 1s as well. it seems like App Engine is the reason. Not sure what configuration on GAE caused this though.
>
>Is there any other way to further lower the Spanner latency? Still have no idea how would it possible to reach 3ms for the read query like the console says. I'm just using 1 spanner node for development now but I doubt that has effect on latency.

I am just doing a drive-by chime in. Would you mind running the experiments after doing a dummy query e.g. perform any sort of read using the client and after that, then do the timing? I ask because some months ago encountered this kind of issue with "cold starts" whereby CreateSession would take 400+ms but after that had gone through, the latencies were in the 20ms to 4ms range. This accompanied me writing about extracting observability signals using OpenCensus in some of these libraries like the Spanner clients for Go and Java as I did documented in https://medium.com/@orijtech/cloud-spanner-instrumented-by-opencensus-and-exported-to-stackdriver-6ed61ed6ab4e

#### Cold start
![spanner-cold-start](https://user-images.githubusercontent.com/4898263/48756977-16459d80-ec50-11e8-9c97-6a45c033e3a2.png)

#### Warm usage
![spanner-warm-start](https://user-images.githubusercontent.com/4898263/48756993-1c3b7e80-ec50-11e8-95d8-c3dc24c2ff9f.png)

#### Typical usage
![spanner-typical](https://user-images.githubusercontent.com/4898263/48757002-23fb2300-ec50-11e8-852a-264cfa1d88a0.png)
